luakit.1.in
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\" Luakit man page.
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.ds appname Luakit
.ds cmdname luakit
.ds manname LUAKIT
.ds version LUAKITVERSION
.ds year 2012
.ds date \*[year]-10-10
.ds appauthors Mason Larobina <mason.larobina@gmail.com>
.ds manauthors Pierre Neidhardt <ambrevar@gmail.com>
.
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.TH \*[manname] 1 "\*[date]" "\*[appname] \*[version]" "User Commands"
.
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH NAME
\*[appname] - Fast, small, WebKit based browser framework extensible by Lua.
.
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH SYNOPSIS
.
.SY \*[cmdname]
.RI [ OPTION ] " " [ URI ]
.YS
.
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH DESCRIPTION
\*[appname] is a highly configurable, browser framework based on the WebKit web
content engine and the GTK+ toolkit. It is very fast, extensible by Lua and
licensed under the GNU GPLv3 license. It is primarily targeted at power users,
developers and any people with too much time on their hands who want to have
fine-grained control over their web browsers behaviour and interface.
.
.SS Modes
\*[appname] can run in various modes, which specify how the user interacts with
the browser.
.TP
.B Normal mode
This is the default mode when \*[appname] is started. You can load URIs, open
tabs and windows, access to other modes, etc.
.TP
.B All mode
Special meta-mode in which the bindings are present in all modes.
.TP
.B Insert mode
When selecting form fields \*[appname] will enter the insert mode which allows
you to enter text in form fields without accidentally triggering normal mode
bindings.
.TP
.B  Passthrough mode
\*[appname] will pass every key event to the WebView until the user presses
Escape. This is useful for using webpage shortcuts.
.TP
.B Command mode
Enter commands. Every action in \*[appname] is a command, so basically you can
do anything from there.
.TP
.B Lua mode
Execute arbitrary Lua commands within the \*[appname] environment.
.
.SS Userscripts
\*[appname] is highly extensible with userscripts written in Lua. These scripts
may provide additional modes. Some are embedded by default. You will need to
load the userscript to use its features. See the \fBCONFIGURATION\fR section for
more details.
.
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH OPTIONS
.SS Help options
.TP
.BR -h ", " --help
Show help options.
.TP
.BR --help-all
Show all help options.
.TP
.BR --help-gtk
Show GTK+ help options.
.
.SS Application options
.TP
.BR -k ", " --check
Check configuration file and exit.
.TP
.BR -c ", " --config = \fIFILE\fR
Configuration file to use.
.TP
.BR -n ", " --nonblock
Fork \*[appname] into the background.
.TP
.BR -U ", " --nounique
Ignore libunique bindings.
.TP
.BR -u ", " --uri = \fIURI\fR
URI(s) to load at startup.
.TP
.BR -v ", " --verbose
Print debugging output.
.TP
.BR -V ", " --version
Print version and exit.
.TP
.BR --display = \fIDISPLAY\fR
X display to use.
.
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH KEYBINDINGS
\*[appname] is fully usable with keyboard. Default bindings should be familiar
to users of Emacs and Vim; nonetheless a lot of keyboards and mouse shortcuts
are shared with popular web browsers.
.P
Pressing
.B :
will open the command prompt. You can enter commands from there.
You can view the current list of bindings
in \*[appname]'s help page with the
.IP
.B :help
.P
command.
.P
Some bindings have an uppercase version which will behave somewhat
differently. Some come with userscripts, so you will need to have the
appropriate userscript loaded to use it.  Here follows a list of some noteworthy
default bindings.
.
.TP
.B ZZ
Quit and save the session.
.TP
.B ZQ
Quit without saving the session.
.
.TP
.BR o ", " O
Open specified URIs. If uppercase, edit current URI.
.TP
.BR t ", " T
Open specified URIs in new tab. If uppercase, edit current URI.
.TP
.BR w ", " W
Open specified URIs in new window. If uppercase, edit current URI.
.P
.BR <Left> ", " <Down> ", " <Up> ", " <Right>
.br
.BR h ", " j ", " k ", " l
.RS 7
Scroll page vertically and horizontally.
.RE
.P
.BR <Home> ", " <End>
.br
.BR gg ", " G
.RS 7
Go to top / bottom of the page.
.RE
.TP
.BR [count]%
Go to [count] percent of the page.
.TP
.BR f ", " F
Enter follow mode. Use numbers or text to open
corresponding links. Use arrow to navigate between links. If uppercase,
open in new tab.
Requires
.I follow
userscript.
.TP
.BR i
Enter insert mode. In some case form fields may not receive characters; press an
arrow key to insert characters correctly.
.TP
.BR gi
Enter insert mode in the first form field.
Requires
.I go_input
userscript.
.TP
.BR <F11>
Toggle fullscreen.
.TP
.BR + ", " -
Change the zoom level.
.TP
.BR =
Restore zoom level.
.TP
.BR p ", " P
Open URI from clipboard. If uppercase, open in new tab.
.TP
.BR y
Yank current URI to clipboard.
.TP
.BR <Ctrl>x ", " <Ctrl>a
Decrement / increment last number in URI. This is useful for forum threads or
any ordered website.
.TP
.BR H ", " L
Go back / forward in the browser history.
.P
.BR <Ctrl><Shift><Tab> ", " <Ctrl><Tab>
.br
.BR <Ctrl><PageDown> ", " <Ctrl><PageUp>
.br
.BR gT ", " gt
.RS 7
Go to previous / next tab.
.RE
.TP
.BR <Alt>[0-9]
Go to tab #, where # is between 0 and 9.
.TP
.BR < ", " >
Reorder tabs.
Requires
.I taborder
userscript.
.TP
.BR [count]d
Close [count] tabs.
.TP
.BR u
Restore last closed tab.
.TP
.BR gy
Duplicate current tab.
.TP
.BR gh ", " gH
Open homepage. If uppercase, open in new tab.
.TP
.BR r ", " R
Reload current page. If uppercase, skip cache.
.TP
.BR <Ctrl>c
Stop loading the current page.
.TP
.BR <Ctrl>z
Enter passthrough mode. Use ESC to return to normal mode.
.TP
.BR M[a-zA-Z0-9]
Associate current URI to quickmark #, where # is an ASCII letter or a digit.
Requires
.I quickmarks
userscript.
.TP
.BR go[a-zA-Z0-9] ", " gn[a-zA-Z0-9] ", " gw[a-zA-Z0-9]
Open specified quickmark in current tab /
new tab / new window.
Requires
.I quickmarks
userscript.
.TP
.BR B
Add current URI to bookmarks.
.TP
.BR gb ", " gB
Open bookmarks manager. If uppercase, open in new tab.
.TP
.BR / ", " ?
Search / reverse search for a string on current page.
.TP
.BR n ", " N
Find next / previous result from search.
.
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH COMMANDS
All bindings actually refer to commands. However, some commands do not have
bindings associated by default. If
.I completion
userscript is used, you can autocomplete commands by pressing <TAB> by
default. Once again, you should consult the configuration files to get an
exhaustive list. The currently available functions may be displayed from the
help view, which you can open with the
.B :help
command.
.P
Here follows some noteworthy commands:
.TP
.B :bookmarks
Display and search bookmarks.
.TP
.B :downloads
Open Download page, which displays all downloads along with their status.
.TP
.B :dump
Download current page.
.TP
.B :save
Save the complete page as a single mhtml file.
.TP
.B :help
Display all commands and bindings, along with their description. The help page
also features some details about modes.
.TP
.B :history
Display and search history.
.TP
.B :inspect
Launch WebKit inspector.
Use \fB:inspect!\fR to toggle off.
.TP
.B :nohlsearch
Disable search highlighting.
.TP
.B :qmarks
Display the quickmarks list.
.TP
.B :tabhistory
Display the tab history.
.TP
.B :view-source
Display page source code.
Use \fB:view-source!\fR to toggle off.
.
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH FILES
.\" TODO: fix indentation.
\*[appname] will load configuration files from the following folders in priority
order:
.IP
.I $XDG_CONFIG_HOME/luakit
.IP
.I $XDG_CONFIG_DIRS/luakit
.P
Default configuration files:
.TP
.I binds.lua
The keyboard and mouse bindings.
.TP
.I globals.lua
General configuration, like home page, search engines, user agent, per-domain
properties, cookies policy.
.TP
.I modes.lua
All default modes are set in this file.
.TP
.I rc.lua
Main configuration files. All other files are loaded from there.
.TP
.I theme.lua,
Colors definition.
.TP
.I webview.lua
WebKit related.
.TP
.I window.lua
Status bar, windows and tabs behaviour.
.P
Embedded userscript will be loaded from \fI$XDG_DATA_DIRS/luakit/lib\fR.
.P
All browsing-related files are stored in \fI$XDG_DATA_HOME/luakit\fR.  All of
are created if needed and if they do not exist. Depending on the userscripts you are using,
you may find:
.TP
.I bookmarks.db
An SQLite3 database containing your complete bookmark list. If you want to
synchronise your bookmarks between your different systems, just share this
file. You can use a symbolic link if the file is not stored in the required
folder.
.TP
.I cookies.db
This file contains all details needed for websites keeping track of your status,
like login information. Delete this file once you are finished with browsing if
you are not on your personal system.
.TP
.I history.db
Your browsing history.  Delete this file once you are finished with browsing if
you are not on your personal system.
.TP
.I quickmarks
A plain text file saving your quickmarks. The structure is extremely simple:
each line is a single quickmark; first character is the quickmark shortcut and
may be one of [a-zA-Z0-9]; second character is a space; the remaining part is
the URI.
.
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH CONFIGURATION
If you do not want to stick to the default configuration and want to benefit
from the flexibility and extensibility of \*[appname], you can fully configure
it from the Lua configuration files.
.P
If you do not want to start from scratch, you may use default configuration
files and tweak them to fit your needs. Use the following command to copy
configuration file to your home folder:
.IP
.EX
cp -r $XDG_CONFIG_DIRS/luakit $XDG_CONFIG_HOME
.EE
.P
You can now edit the new files. Configuration should be quite straightforward,
even if you do not know much about Lua.
.
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH AUTHORS
\*[appname] was mainly developed by \*[appauthors]. Other contributors are
listed in the
.I AUTHORS
file.
.
.P
This man page was written by \*[manauthors].


CHANGELOG.md
# Changelog

## [2.0/2.1]

### Migrating from version 2017-08-10

 1. Remove the two `if unique then ... end` blocks from your `rc.lua`.
 2. Add `require "unique_instance"` to your `rc.lua`, before all other `require` statements.
 4. Remove all configuration files except `rc.lua` and `theme.lua`. Any changes to `globals.lua`
    need to be migrated to `rc.lua` and changed to use the `settings` API.

### Added

 - Added `styles.new_style` function.
 - Added `styles.toggle_sheet` function.
 - Added `styles.watch_styles` function, and enabled live-editing of user styles.
 - Added `luakit.install_paths` table. `luakit.install_path` is now deprecated.
 - Added `Control-Y` readline binding.
 - Added ability to control whether links from secondary instances open in a new window.
 - Added `luakit.resource_path` property to control where luakit searches for resource files.
 - Added `lousy.util.find_resource` function.
 - Added `scroll` signal.
 - Added ability to bind actions to webview scroll events.
 - Added ability to set the default zoom level.
 - Added `webview` widget `"permission-request"` signal.
 - Added `webview` widget `hardware_acceleration_policy` property.
 - Added `webview` widget `allow_file_access_from_file_urls` and `allow_universal_access_from_file_urls` properties.
 - Added `settings` module and APIs. This replaces the `domain_props` module.
 - Added `tablist.always_visible` setting.
 - Added `utf8.len` (same as `string.wlen`) and `utf8.offset` methods.
 - Added `utf8.charpattern` property.
 - Added `:set` and `:seton` commands, for changing settings.
 - Added ability to always save session before exiting luakit.
 - Added `markup` option to window `set_prompt()` method.
 - Added `detach-tab` signal.
 - Added support for multi-byte characters in hints.
 - Added widget `replace` method.

### Changed

 - It is no longer necessary to add bindings to tables with `lousy.bind.add_binds()`.
 - Readline bindings have been moved to `readline.lua`.
 - Readline bindings are now automatically bound when the input bar is visible.
 - Unique instance support has been moved to `unique_instance.lua`.
 - The `image` widget now uses `luakit.resource_path` to locate local files.
 - The log viewer now shows errors logged by a user-defined rc.lua failing to load.
 - Luakit will now remove its IPC socket file before restarting.
 - The editor command now defaults to using `xdg-open` to edit files.  The `default` builtin
   command has been renamed `autodetect`.
 - Changed `luakit://introspector/` to `luakit://binds/`.
 - URL completion now uses word-based fuzzy matching.
 - `:download` now uses the current page URI by default.
 - `gy` now accepts a count.
 - `:tabopen` will now only open local files when given an absolute path.
 - `:styles-list` now lists active styles first and disabled styles last.

### Removed

 - Removed `domain_props` module. It is replaced by the `settings` module and its APIs.
 - Removed all configuration files except `rc.lua` and `theme.lua`.
 - Removed `enable_private_browsing` webview property.
 - Removed `w.closed_tabs` field. It is now private to the `undoclose` module.

### Fixed

 - Fixed <luakit://help/> not finding documentation with custom DOCDIR
 - Various minor documentation fixes.
 - Fixed `Control-Scroll` and `Shift-Scroll` key bindings not working with smooth scrolling.
 - Fixed inability to switch focus between web page elements with `Tab` and `Shift-Tab`.
 - Fixed log page bug when logging messages with newlines.
 - Fixed `Up` and `Down` keybindings being broken on completion menu.
 - Fixed hardcoded path to luakit icon.
 - Fixed luakit:// pages not working and spewing errors when not using LuaJIT.
 - Fixed thumbnail hinting not retrieving thumbnail links correctly.
 - Fixed inability to bind `Modifier-Minus`.
 - Fixed readline handling of wide characters.
 - Fixed completion not suggesting history/bookmarks items without titles/tags.
 - Fixed `:dump` command not working due to use of a removed API.
 - Fixed follow hints being sometimes truncated by the viewport edge.
 - Follow mode now renders hints much faster.
 - Fixed Forward/Back keys not working due to outdated bind syntax.
 - Fixed opening local files with names containing spaces from `:tabopen` and the command-line.

## [2017-08-10]

 - Required WebKitGTK+ version: 2.16+

### Breaking changes

 - Support for WebKitGTK+ versions older than 2.16 has been removed.
 - It is no longer possible to override built-in luakit modules with Lua
   files in one's personal configuration directory.
 - The configuration files `binds.lua` and `modes.lua` have become built-in luakit modules.
   Configuration files named `binds.lua` or `modes.lua` will not be loaded. Any custom
   bindings should be moved to `rc.lua`.

### Added

 - New `history.frozen` API allows temporarily freezing history collection.
 - New `lousy.widget.zoom` statusbar widget: shows current page zoom level.
 - Added `log` signal, emitted whenever a message is logged.
 - New `widget.is_alive` property. Can be accessed even if the widget has been destroyed.
 - Added `luakit://log/` chrome page: displays log messages.
 - Added status bar widget that notifies of any Lua warnings or errors.
 - Added migration, quick-start, and files and directories guides to documentation.
 - Added frequently-asked questions to documentation.
 - Added `luakit.wch_upper` and `luakit.wch_lower` key case conversion utility functions.
 - Added `formfiller.extend` function for extending the formfiller DSL.
 - Added context-aware command completion.

### Fixed

 - Fixed code-blocks in documentation being formatted incorrectly.
 - Fixed incompatibility of `editor.lua` with urxvt.
 - Fixed slow performance while beginning a search.
 - Fixed `lousy.util.table.join` merging tables in unpredictable order.
 - Fixed `image_css` raising errors on page zoom in/out.
 - Worked around `image_css` breaking slightly when using non-1.0 zoom_level.

luakit.c
/*
 * luakit.c - luakit main functions
 *
 * Copyright © 2010-2011 Mason Larobina <mason.larobina@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include "common/util.h"
#include "globalconf.h"
#include "luah.h"
#include "ipc.h"
#include "log.h"
#include "web_context.h"

#include <errno.h>
#include <gtk/gtk.h>
#include <locale.h>
#include <signal.h>
#include <stdlib.h>
#include <sys/wait.h>
#include <unistd.h>
#include <webkit2/webkit2.h>

#if !WEBKIT_CHECK_VERSION(2,16,0)
#error Your version of WebKit is outdated!
#endif

static void
init_directories(void)
{
    /* create luakit directory */
    globalconf.cache_dir  = g_build_filename(g_get_user_cache_dir(),  "luakit", globalconf.profile, NULL);
    globalconf.config_dir = g_build_filename(g_get_user_config_dir(), "luakit", globalconf.profile, NULL);
    globalconf.data_dir   = g_build_filename(g_get_user_data_dir(),   "luakit", globalconf.profile, NULL);
    g_mkdir_with_parents(globalconf.cache_dir,  0700);
    g_mkdir_with_parents(globalconf.config_dir, 0700);
    g_mkdir_with_parents(globalconf.data_dir,   0700);
}

static void
parse_log_level_option(gchar *log_lvl)
{
    gchar **parts = g_strsplit(log_lvl, ",", 0);

    for (gchar **part = parts; *part; part++) {
        log_level_t lvl;
        if (!log_level_from_string(&lvl, *part))
            log_set_verbosity("all", lvl);
        else {
            gchar *sep = strchr(*part, '=');
            if (sep && !log_level_from_string(&lvl, sep+1)) {
                *sep = '\0';
                log_set_verbosity(*part, lvl);
            } else
                warn("ignoring unrecognized --log option '%s'", *part);
        }
    }
    g_strfreev(parts);
}

/* load command line options into luakit and return uris to load */
static gchar **
parseopts(int *argc, gchar *argv[], gboolean **nonblock)
{
    GOptionContext *context;
    gboolean *version_only = NULL;
    gboolean *check_only = NULL;
    gchar **uris = NULL;
    globalconf.profile = NULL;
    gboolean verbose = FALSE;
    gchar *log_lvl = NULL;

    /* save luakit exec path */
    globalconf.execpath = g_strdup(argv[0]);
    globalconf.nounique = FALSE;

    /* define command line options */
    const GOptionEntry entries[] = {
        { "check",    'k', 0, G_OPTION_ARG_NONE,         &check_only,          "check config and exit",     NULL   },
        { "config",   'c', 0, G_OPTION_ARG_STRING,       &globalconf.confpath, "configuration file to use", "FILE" },
        { "profile",  'p', 0, G_OPTION_ARG_STRING,       &globalconf.profile,  "profile name to use",       "NAME" },
        { "nonblock", 'n', 0, G_OPTION_ARG_NONE,         nonblock,             "run in background",         NULL   },
        { "nounique", 'U', 0, G_OPTION_ARG_NONE,         &globalconf.nounique, "ignore libunique bindings", NULL   },
        { "uri",      'u', 0, G_OPTION_ARG_STRING_ARRAY, &uris,                "uri(s) to load at startup", "URI"  },
        { "verbose",  'v', 0, G_OPTION_ARG_NONE,         &verbose,             "print verbose output",      NULL   },
        { "log",      'l', 0, G_OPTION_ARG_STRING,       &log_lvl,             "specify precise log level", "NAME" },
        { "version",  'V', 0, G_OPTION_ARG_NONE,         &version_only,        "print version and exit",    NULL   },
        { NULL,       0,   0, 0,                         NULL,                 NULL,                        NULL   },
    };

    /* Save a copy of argv */
    globalconf.argv = g_ptr_array_new_with_free_func(g_free);
    for (gint i = 0; i < *argc; ++i)
        g_ptr_array_add(globalconf.argv, g_strdup(argv[i]));

    /* parse command line options */
    context = g_option_context_new("[URI...]");
    g_option_context_add_main_entries(context, entries, NULL);
    g_option_context_add_group(context, gtk_get_option_group(FALSE));
    g_option_context_parse(context, argc, &argv, NULL);
    g_option_context_free(context);

    /* Trim unparsed arguments off copy of argv */
    for (gint i = 0; i < *argc; ++i) {
        while ((unsigned)i < globalconf.argv->len && !strcmp(g_ptr_array_index(globalconf.argv, i), argv[i]))
            g_ptr_array_remove_index(globalconf.argv, i);
    }

    /* print version and exit */
    if (version_only) {
        g_printf("luakit %s\n", VERSION);
        g_printf("  built with webkit %i.%i.%i ", WEBKIT_MAJOR_VERSION, WEBKIT_MINOR_VERSION, WEBKIT_MICRO_VERSION);
        g_printf("(installed version: %u.%u.%u)\n", webkit_get_major_version(), webkit_get_minor_version(), webkit_get_micro_version());
        exit(EXIT_SUCCESS);
    }

    if (!log_lvl)
        log_set_verbosity("all", verbose ? LOG_LEVEL_verbose : LOG_LEVEL_info);
    else {
        log_set_verbosity("all", LOG_LEVEL_info);
        parse_log_level_option(log_lvl);
        if (verbose)
            warn("invalid mix of -v and -l, ignoring -v...");
    }

    /* check config syntax and exit */
    if (check_only) {
        init_directories();
        luaH_init(NULL);
        if (!luaH_parserc(globalconf.confpath, FALSE)) {
            g_fprintf(stderr, "Confiuration file syntax error.\n");
            exit(EXIT_FAILURE);
        } else {
            g_fprintf(stderr, "Configuration file syntax OK.\n");
            exit(EXIT_SUCCESS);
        }
    }

    if (uris && argv[1])
        fatal("invalid mix of -u and default uri arguments");

    if (uris)
        return uris;
    else
        return g_strdupv(argv + 1);
}

#if __GLIBC__ == 2 && __GLIBC_MINOR__ >= 50
static GLogWriterOutput
glib_log_writer(GLogLevelFlags log_level_flags, const GLogField *fields, gsize n_fields, gpointer UNUSED(user_data))
{
    const gchar *log_domain = "(unknown)",
                *message = "(empty)",
                *code_file = "(unknown)",
                *code_line = "(unknown)";

    for (gsize i = 0; i < n_fields; ++i) {
        if (!strcmp(fields[i].key, "GLIB_DOMAIN")) log_domain = fields[i].value;
        if (!strcmp(fields[i].key, "MESSAGE")) message = fields[i].value;
        if (!strcmp(fields[i].key, "CODE_FILE")) code_file = fields[i].value;
        if (!strcmp(fields[i].key, "CODE_LINE")) code_line = fields[i].value;
    }

    /* Probably not necessary, but just in case... */
    if (!(G_LOG_LEVEL_MASK & log_level_flags))
        return G_LOG_WRITER_UNHANDLED;

    log_level_t log_level = ((log_level_t[]){
        [G_LOG_LEVEL_ERROR]    = LOG_LEVEL_fatal,
        [G_LOG_LEVEL_CRITICAL] = LOG_LEVEL_warn,
        [G_LOG_LEVEL_WARNING]  = LOG_LEVEL_warn,
        [G_LOG_LEVEL_MESSAGE]  = LOG_LEVEL_info,
        [G_LOG_LEVEL_INFO]     = LOG_LEVEL_verbose,
        [G_LOG_LEVEL_DEBUG]    = LOG_LEVEL_debug,
    })[log_level_flags];

    _log(log_level, code_line, code_file, "%s: %s", log_domain, message);
    return G_LOG_WRITER_HANDLED;
}
#endif

gint
main(gint argc, gchar *argv[])
{
    gboolean *nonblock = NULL;
    globalconf.starttime = l_time();

    log_init();

    /* set numeric locale to C (required for compatibility with
       LuaJIT and luakit scripts) */
    gtk_disable_setlocale();
    setlocale(LC_ALL, "");
    setlocale(LC_NUMERIC, "C");

    /* parse command line opts and get uris to load */
    gchar **uris = parseopts(&argc, argv, &nonblock);

    /* hide command line parameters so process lists don't leak (possibly
       confidential) URLs */
    for (gint i = 1; i < argc; i++)
        memset(argv[i], 0, strlen(argv[i]));

    globalconf.windows = g_ptr_array_new();

    /* if non block mode - respawn, detach and continue in child */
    if (nonblock) {
        pid_t pid = fork();
        if (pid < 0) {
            fatal("Cannot fork: %d", errno);
        } else if (pid > 0) {
            exit(EXIT_SUCCESS);
        }
        pid_t sid = setsid();
        if (sid < 0) {
            fatal("New SID creation failure: %d", errno);
        }
    }

    gtk_init(&argc, &argv);

#if __GLIBC__ == 2 && __GLIBC_MINOR__ >= 50
    g_log_set_writer_func(glib_log_writer, NULL, NULL);
#endif
    init_directories();
    web_context_init();
    ipc_init();
    luaH_init(uris);

    /* parse and run configuration file */
    if (!luaH_parserc(globalconf.confpath, TRUE))
        fatal("couldn't find rc file");

    if (!globalconf.windows->len)
        fatal("no windows spawned by rc file, exiting");

    gtk_main();
    return EXIT_SUCCESS;
}

// vim: ft=c:et:sw=4:ts=8:sts=4:tw=80
